catkin_tools/verbs/catkin_build/build.py

```diff
@@ -195,13 +189,18 @@ def determine_packages_to_be_built(packages, context):
                 pkg_deps = get_cached_recursive_build_depends_in_workspace(package, ordered_packages)
                 packages_to_be_built_deps.extend(pkg_deps)
     else:
-        packages_to_be_built = ordered_packages
+        # Only use whitelist when no other packages are specified
+        if len(context.whitelist) > 0:
+            packages_to_be_built = [p for p in ordered_packages if (p[1].name in context.whitelist)]
+        else:
+            packages_to_be_built = ordered_packages
 
     # Filter packages with blacklist
     if len(context.blacklist) > 0:
-        packages_to_be_built = [p for p in packages_to_be_built if (p[0] not in context.blacklist or p[0] in packages)]
+        packages_to_be_built = [
+            p for p in packages_to_be_built if (p[1].name not in context.blacklist or p[1].name in packages)]
         packages_to_be_built_deps = [
-            p for p in packages_to_be_built_deps if (p[0] not in context.blacklist or p[0] in packages)]
+            p for p in packages_to_be_built_deps if (p[1].name not in context.blacklist or p[1].name in packages)]
```

Comment:
It would probably be better to expand the tuple rather than access it by index. Something like:

```python
[(pth, pkg) for pth, pkg in packages_to_be_built_deps if pkg.name not in context.blacklist or pkg.name in packages]
```